src/python/pants/option/global_options.py

```diff
@@ -59,6 +60,14 @@ def to_glob_match_error_behavior(self) -> GlobMatchErrorBehavior:
         return GlobMatchErrorBehavior(self.value)
 
 
+@dataclass(frozen=True)
+class AuthPluginResult:
+    """The return type for a function specified via `--remote-auth-plugin`."""
+
+    store_headers: Dict[str, str]
+    execution_headers: Dict[str, str]
```

Comment:
Originally, I was including two extra fields: `expiration` and `host_name`. But I removed this because this plugin is severely limited in what it can do and when it can execute. Imo, it's premature to include those two fields because we cannot reasonably consume the data anytime soon.

It's easier to add it in the future than to remove it.

Reply:
it is technically not frozen if members are dict.
so either we remove the frozen or we make the headers a immutable data type, i.e. something like a `Tuple[Tuple[str,str],...]`
also, it is not clear why there are two fields for headers, there should be one.
we can get rid of the `execution_headers` because we are not going to do remote execution anytime soon.
and when we do, from an auth POV, the same headers (which will contain auth data) can be used for both.

I also, I don't see how adding fields that the caller ignores limits the plugin, you can just define them like so:
```python
expiration: Optiona[datetime] = None
host: Optional[str] = None
```
and no one has to really care about them until some point in the future.

also missing from this data class is some kind of state enum that will indicate if remote cache should even be used.
i.e. 
```python
@unqiue
class AuthState(Enum):
  OK = "ok"
  UNAVAILABLE = "unavailable"
```

in the future there will be more values (like TRANSIENT_FAILURE) which will let the caller know that remote caching is not available right now but it is ok to try again in a few seconds.

Reply:
> also, it is not clear why there are two fields for headers, there should be one.
we can get rid of the execution_headers because we are not going to do remote execution anytime soon.

Because this corresponds to `--remote-store-headers` and `--remote-execution-headers`. They are currently two distinct things, both on the options system and in the Rust code. We must reflect that here. It's okay that often they're the same thing.

> I also, I don't see how adding fields that the caller ignores limits the plugin

I think there's some miscommunication. I am not saying that we're limiting the plugin author.

I'm saying that the current plugin architecture is extremely limited. It's in an awkward place and can't do many things like use the Rules API or be re-executed during the same Pants run. So, until we re-architect this entire thing (possibly replacing it all), we wouldn't be able to do anything with those two fields in the Pants client.

I don't see any benefit to adding two fields that cannot be used by the Pants client without a big re-architecture. At that point, we can easily add the fields then, based on the new design, no? It's easier to add code than to remove it.

> also missing from this data class is some kind of state enum that will indicate if remote cache should even be used.

Interesting. that's outside the scope of what this plugin currently does - right now, it's purely for setting the headers. But I think that sounds sensible for it to be able to turn off remote caching / execution entirely if no token was available. That's technically feasible to implement, too.

@tdyas any thoughts?

Reply:
as discussed in the last meeting, disabling remote cache by the auth plugin (when there is no token available) needs to supported in the initial version of this.
This is a part of the current milestone.

Reply:
as for the the `execution_headers` there is absolutely no need to include it as part of the interface for this plugin.
there is no remote execution in the next happening in near future. so having this param is pointless and might actually break things when/if we try to work on remote execution.

Reply:
as for the missing fields (host, expiration) is is likely that when we work on a long term solution (i.e. returning this dataclass from a rule) we would want to use the same dataclass, instead of defining a new one. so having this class contains those fields makes a lot of sense even now.
Also, overriding the host and needing to respect that might be required sooner rather than later (due to latency consideration when integrating with other repos).

Reply:
> as discussed in the last meeting, disabling remote cache by the auth plugin (when there is no token available) needs to supported in the initial version of this.

Okay, I'll work on something.

> as for the the execution_headers there is absolutely no need to include it as part of the interface for this plugin.
there is no remote execution in the next happening in near future. so having this param is pointless and might actually break things when/if we try to work on remote execution.

If you don't want to touch execution_headers at all, then simply return the `initial_execution_headers` kwarg unmodified. Or, return the exact same thing as the modified `store_headers`. Most the time, those end up being the same.

This is a generic plugin interface to set headers for both types of remoting; we can't be leaving off the ability to change execution_headers simply because Toolchain is not at the moment using remote execution.

Reply:
> is is likely that when we work on a long term solution (i.e. returning this dataclass from a rule) we would want to use the same dataclass, instead of defining a new one. so having this class contains those fields makes a lot of sense even now.

I don't understand that reasoning - we can add those fields then, when it's needed, right? We can set default values for those new fields so that we aren't breaking any previously written plugins.

For now, this would be prematurely writing code for a speculative future, especially given that the current plugin architecture doesn't allow the client to do anything interesting with those fields.

Reply:
> we can get rid of the execution_headers because we are not going to do remote execution anytime soon.

Toolchain isn't using it, but there is at least one Pants user using remote execution that I am aware of. There could be others. Remote execution is still a documented feature of Pants.

Reply:
I doubt the need to provide the headers using this kind of plugin.

Reply:
When we do have remote execution, the plugin would need to put the auth header into both `store_headers` and `execution_headers` for remote execution auth to work as expected. That said, if we are including anticipated items in this interface, that is reason enough to include it.